pkg/ingester/ingester.go

```diff
@@ -506,6 +508,12 @@ func (i *Ingester) stopping(_ error) error {
 	}
 	i.flushQueuesDone.Wait()
 
+	// In case the flag to clear tokens on shutdown is set we need to mark the
+	// ingester service as "failed", so Loki will shut down entirely.
+	// The module manager logs the failure `modules.ErrStopProcess` in a special way.
+	if i.lifecycler.ClearTokensOnShutdown() && errs.Err() == nil {
+		return modules.ErrStopProcess
```

Comment:
Why can't we just rely on `nil` error? We can return internal server error from `ShutdownAndForgetHandler` handler when error is non-nil or am I missing something?

Reply:
No, because we need to be able to distinguish between stopping of the ingester service that happened through the signal handler, and stopping of the service that was triggered by the shutdown handler. In case of the former, the error will be `nil` as well.

Reply:
In both cases, we are returning `w.WriteHeader(http.StatusNoContent)` so I don't see any special handling in `ShutdownAndForgetHandler` for either of the cases. Am I missing something?